### PR TITLE
MM-52804 - Implement SendPushNotification plugin api method

### PR DESCRIPTION
**Pull request by cpoile**
#### Summary
- The intention of this method is to allow plugins to send their own push notifications. It uses the `model.PluginPushNotification` struct added in #24263 which is what's needed to send a push notification to all connected sessions for UserID.
- Note: Plugins will be constrained by the existing `getPushNotificationMessage` rules, i.e. we can't override the server's `cfg.EmailSettings.PushNotificationContents` setting `id_loaded` setting. But if it is set to `full`, we will still be constrained to a push notification that reads: `@SenderName: post message` (for GMs). The designs for Calls notifications have the message as: `SenderName is inviting you to a call`, but this constraint will turn that into: `@SenderName: is inviting you to a call` (for GMs). For DMs it sends two lines, like: `SenderName <next line> post.Message` without additional formatting.
- I'm considering adding a new flag to `app.sendPushNotification`, something like `noNameSeparator`, so we can have: `@SenderName is inviting you to a call` for GMs and `SenderName <next line> is inviting you to a call` for DMs. Would appreciate thoughts on it.
- Companion to #24263 

#### Ticket Link
- https://mattermost.atlassian.net/browse/MM-53924

#### Release Note
```release-note
Added a `NotificationWillBePushed` plugin api method which allows plugins to send push notifications to a specific user's mobile sessions.
```

### CODE REVIEW

File: server/channels/app/plugin_api.go [hunk 1]
```diff
@@ -1262,3 +1263,34 @@ func (api *PluginAPI) GetUploadSession(uploadID string) (*model.UploadSession, e
 	}
 	return fi, nil
 }
+
+// SendPushNotification will attempt to send a push notification to `notification.User`, using
+// `notification.Post` as the source of the notification. Note: the NotificationWillBePushed hook will
+// be run after SendPushNotification is called.
+func (api *PluginAPI) SendPushNotification(notification *model.PluginPushNotification) error {
```
Comment (lieut-data) [inline server/channels/app/plugin_api.go:1270]: This adds it to the `app` side of the interface, but I don't see the corresponding changes to the RPC layer and interface that make it actually work with a real plugin. Is that somehow separate from this PR?

File: server/channels/app/plugin_api.go [hunk 2]
```diff
@@ -1262,3 +1263,34 @@ func (api *PluginAPI) GetUploadSession(uploadID string) (*model.UploadSession, e
...
+		return appErr
+	}
+
+	postNotification := &PostNotification{
+		Post:       notification.Post,
+		Channel:    notification.Channel,
+		ProfileMap: profiles,
+		Sender:     sender,
+	}
+	api.app.sendPushNotification(postNotification, user, false, false, "")
```
Comment (lieut-data) [inline server/channels/app/plugin_api.go:1294]: I don't know what `false`, `false`, and `""` are, but also wondering if plugins might want to customize these or other parameters? See comments in parent PR, but wondering if we can just accept the full `model.PostNotification` here and backfill anything in that the plugin leaves unset?

**Review by lieut-data — CHANGES_REQUESTED** [review]
Thanks, @cpoile! Two quick thoughts for a first pass.

File: server/channels/app/plugin_api.go (hunk 2 above)
Comment (cpoile) [reply server/channels/app/plugin_api.go:1291]: Those are `explicitMention`, `channelWideMention`, and `replyToThreadType` -- first two are for generating the default messages when the server's `cfg.EmailSettings.PushNotificationContents` is set to certain values (or the server isn't fully licensed), the last is for CRT thread replies. We could definitely accept those as well in the `model.PluginPushNotification`. Let me make that change.

I'm not against accepting the `model.PostNotification`, as it would simplify a lot of the contextual knowledge you need when figuring out how to make a push notification from a plugin. But I decided against it because of a couple problems I encountered:
1. Our notification system assumes that all push notifications refer to a post. So if we accept a raw `model.PostNotification` we would be putting it on the plugin to make sure the ids are all set correctly. This way, the plugin just has to know "valid Post", "valid Channel", and a target userID, and leaves the rest to the system (including following the server's rules for the message, given licensing and cfg settings).
2. The similar problem to https://github.com/mattermost/mattermost/pull/24263#discussion_r1297090335 -- the plugin probably shouldn't be allowed to send raw message data if the server has `IdLoadedNotification` set.
 
Basically: The way the PR does it, it allows a plugin to "suggest" a push notification, but the server will process it and make it conform to what is allowed, based on cfg settings and licensing.

Anyway, that's my thinking. Open to alternatives!

Comment (lieut-data) [reply server/channels/app/plugin_api.go]: Thanks, @cpoile! I like the framing of suggesting a push notification as it more clearly captures the intent. If we could find a name that captures this intent (leaving `sendPushNotification` for some future, more powerful API addition), I think my concerns would largely disappear.

File: server/channels/app/notification_push.go [hunk 3]
```diff
@@ -143,9 +143,12 @@ func (a *App) sendPushNotification(notification *PostNotification, user *model.U
...
 		rejectionReason := hooks.NotificationWillBePushed(&model.PluginPushNotification{
-			Post:    notification.Post.ForPlugin(),
-			Channel: notification.Channel,
-			UserID:  user.Id,
+			Post:               notification.Post.ForPlugin(),
+			Channel:            notification.Channel,
+			UserID:             user.Id,
+			ExplicitMention:    explicitMention,
+			ChannelWideMention: channelWideMention,
+			ReplyToThreadType:  replyToThreadType,
```
Comment (lieut-data) [inline server/channels/app/notification_push.go:151]: For the plugin caller, can we document these on `PluginPushNotification` to explain when to set them?

**Review by lieut-data — APPROVED** [review]
Thanks, @cpoile!

File: server/channels/app/plugin_api.go (hunk 2 above)
Comment (cpoile) [reply server/channels/app/plugin_api.go]: As discussed offline, moving to `SendPluginPushNotification` is a better indication for what it does. Thanks!

File: server/channels/app/notification_push.go (hunk 3 above)
Comment (cpoile) [reply server/channels/app/notification_push.go:151]: Done, thanks!

File: server/channels/app/plugin_api.go [hunk 4]
```diff
@@ -1262,3 +1263,31 @@ func (api *PluginAPI) GetUploadSession(uploadID string) (*model.UploadSession, e
 	}
 	return fi, nil
 }
+
+func (api *PluginAPI) SendPluginPushNotification(notification *model.PluginPushNotification) error {
+	var profiles map[string]*model.User
+	var err error
+	if notification.Channel.Type == model.ChannelTypeGroup {
+		if profiles, err = api.app.Srv().Store().User().GetAllProfilesInChannel(context.Background(), notification.Channel.Id, true); err != nil {
```
Comment (streamer45) [inline server/channels/app/plugin_api.go:1271]: Should we pass `api.ctx` context here?

**Review by streamer45 — APPROVED** [review]
Looks good :+1:

File: server/channels/app/plugin_api.go (hunk 4 above)
Comment (cpoile) [reply server/channels/app/plugin_api.go:1271]: Makes sense, done!

**Comment by cpoile** [comment]
Apologies for force-push; wanted to rebase to make an easier merge after base is merged.

**Comment by cpoile** [comment]
Tested and works as intended.
